Cluster features on raft: add storage for supported and enabled features

piodul opened the pull request:
This PR implements the storage part of the cluster features on raft functionality, as described in the "Cluster features on raft v2" doc. These changes will be useful for later PRs that will implement the remaining parts of the feature.

Two new columns are added to `system.topology`:

- `supported_features set<text>` is a new clustering column which holds the features that given node advertises as supported. It will be first initialized when the node joins the cluster, and then updated every time the node reboots and its supported features set changes.
- `enabled_features set<text>` is a new static column which holds the features that are considered enabled by the cluster. Unlike in the current gossip-based implementation the features will not be enabled implicitly when all nodes support a feature, but rather via an explicit action of the topology coordinator.

These columns are reflected in the `topology_state_machine` structure and are populated when the topology state is loaded. Appropriate methods are added to the `topology_mutation_builder` and `topology_node_mutation_builder` in order to allow setting/modifying those columns.

During startup, nodes update their corresponding `supported_features` column to reflect their current feature set. For now it is done unconditionally, but in the future appropriate checks will be added which will prevent nodes from joining / starting their server for group 0 if they can't guarantee that they support all enabled features.

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1802/

gleb-cloudius commented:
Drop when from "but rather when via an explicit action of the topology coordinator."

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -256,11 +256,13 @@ schema_ptr system_keyspace::topology() {
             .with_column("current_cdc_generation_uuid", uuid_type, column_kind::static_column)
             .with_column("current_cdc_generation_timestamp", timestamp_type, column_kind::static_column)
             .with_column("global_topology_request", utf8_type, column_kind::static_column)
+            .with_column("enabled_features", set_type_impl::get_instance(utf8_type, true))
```
Should be static, no?

piodul replied:
Yes, definitely, I'll change it.

piodul replied:
I mentioned it in the cover letter:

> `enabled_features set<text>` is a new static column which holds the features that are considered enabled by the cluster. Unlike in the current gossip-based implementation the features will not be enabled implicitly when all nodes support a feature, but rather via an explicit action of the topology coordinator.

avikivity replied:
Ok. Sorry for not noticing (I'm really good at asking to put stuff in cover letters, but very bad at actually reading them). It makes sense for the enabling to be an explicit step (since some transactional action might be needed, like changing the schema).

gleb-cloudius commented on db/system_keyspace.cc:
```diff
@@ -3651,6 +3669,14 @@ future<service::topology> system_keyspace::load_topology_state() {
+            auto cdef = topology()->get_column_definition("enabled_features");
+            auto deserialized = cdef->type->deserialize(blob);
+            auto features = value_cast<set_type_impl::native_type>(deserialized);
+            ret.enabled_features = decode_features(features);
```
Lets put the code in the function since it is copied twice.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -615,6 +624,28 @@ topology_node_mutation_builder& topology_node_mutation_builder::set(const char*
+        cm.tomb = tombstone(_builder._ts - 1, gc_clock::now());
+        _r.cells().apply(*cdef, cm.serialize(*cdef->type));
+    } else {
+        del(cell);
```
Why do you want to delete by writing an empty set?

piodul replied:
I based this function on the overload of `topology_node_mutation_builder::set` for the token set which does the same thing. Maybe it would make more sense to write an empty set here... but then I think the overload for tokens should be changed as well.

All in all, I don't think this function will be ever called with an empty feature set, so it doesn't matter what it does in this case. Perhaps I can try to unify this overload with the one for setting tokens.

avikivity replied:
It's cleaner to do the conversion at the caller, rather than duplicate this large function for every type.

Also the original set might be sorted incorrectly. So perhaps the template should create a new set<sstring>, and call a non-template function to create the mutation.

Also: I want one day to be able to write `auto m = cql_statement_as_mutation("UPDATE tab SET my_set = ?", a_set);`

piodul replied:
This should be fixed now. The setters for both set of features and set of tokens use an internal helper which encapsulates most of the nitty-gritty details and ensures proper order of the values.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -649,6 +680,23 @@ topology_mutation_builder& topology_mutation_builder::set_global_topology_reques
+
+    _m.partition().static_row().apply(*cdef, cm.serialize(*cdef->type));
+    return *this;
+}
```
I think those two function can mostly shared the code as well. The only serious difference is `row().apply()` vs `static_row().apply()`

piodul replied:
> I think those two function can mostly shared the code as well.

I'll try.

> The only serious difference is `row().apply()` vs `static_row().apply()`

The other important difference is that it adds features instead of overwriting them.

gleb-cloudius reviewed (commented):
Left some comments.

avikivity commented:
Please add explanation in the cover letter about the upgrade strategy.

avikivity commented:
> Please add explanation in the cover letter about the upgrade strategy.

Or maybe it's irrelevant yet?

avikivity commented on service/storage_service.cc:
```diff
@@ -515,6 +520,8 @@ class topology_node_mutation_builder {
     topology_node_mutation_builder& set(const char* cell, const raft::server_id& value);
     topology_node_mutation_builder& set(const char* cell, const std::unordered_set<dht::token>& value);
+    template<ConvertibleToSstring S>
+    topology_node_mutation_builder& set(const char* cell, const std::set<S>& value);
```
std::convertible_to<sstring>

piodul replied:
Right, I forgot it exists...

piodul replied:
Actually, it's not the same. It appears that `std::string_view` is not `std::convertible_to<sstring>`. The overload does not match on `std::set<std::string_view>` and instead it matches on the first overload, which converts the set to `sstring` via `::format`...

It looks like the `sstring` constructor that accepts `std::string_view` is explicit, and `std::convertible_to` does not take it into account.

`std::constructible_from<sstring, S>` seems to do the job.

piodul commented:
> > Please add explanation in the cover letter about the upgrade strategy.
> 
> Or maybe it's irrelevant yet?

For now, cluster features on raft will use the topology coordinator and will be hidden behind the experimental raft feature. We don't have an upgrade strategy for the topology coordinator yet, but AFAIK we will design it in the future and cluster features will be handled as a part of it.

piodul commented:
v2:
- Fixed the column kind for `enabled_features` (it's now a static column, as declared in the commit message)
- Removed "when" from the cover letter
- Abstracted away a common pattern in the code that deserializes sets in system_keyspace.cc to a function
- The topology mutation builder and topology node mutation builder now derive from a common CRTP class which abstracts some details about setting columns in the mutation being built, including proper ordering of values in set columns
- Removed the error prone `topology_node_mutation_builder::set` overload that accepted all types that can be formatted

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/1896/

kostja commented on service/storage_service.cc:
```diff
@@ -499,12 +499,42 @@ future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
    co_await _db.local().apply(freeze(muts), db::no_timeout);
 }
 
+template<typename Builder>
```
Given @kbr-scylla will make the query processor available soon-ish early at startup phase maybe we don't need to invest so much into the mutation builder logic  - we will be switching topology code to CQL. But the patch itself is fine of course.

kbr-scylla replied:
> we will be switching topology code to CQL

Huh?

kostja replied:
We discussed this when the centralized topology coordinator was introduced. Using the mutation builder is there only because we wanted to avoid CQL and wanted something quick.

kbr-scylla replied:
There was nothing preventing us from using CQL and my refactors don't affect it. Topology coordinator code sits in `storage_service` and is executed when `query_processor` is already available.

kostja replied:
I think I remember clearer now, the other problem was that group0 client needs a mutation.

gleb-cloudius replied:
> I think I remember clearer now, the other problem was that group0 client needs a mutation.

This is the only reason .  We prepare raft commands here, not writing into a table.

gleb-cloudius replied:
What we discussed is that we need to optimize loading topology to not re-read everything,  but look into a mutation in a command and figure what needs to be changed.

gleb-cloudius commented on service/storage_service.cc:
```diff
@@ -538,10 +538,8 @@ class topology_node_mutation_builder
-        return set(cell, sstring{::format("{}", value)});
-    }
+    topology_node_mutation_builder& set(const char* cell, node_state value);
+    topology_node_mutation_builder& set(const char* cell, topology_request value);
```
Can't we restrict the T to be one of two types by a concept?

kbr-scylla replied:
Personally I prefer this version.

gleb-cloudius replied:
> Personally I prefer this version.

Just wondering how many duplication it would take for you to prefer concepts. 5, 15, 50? :)

gleb-cloudius reviewed (approved):
Small nit only.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -499,12 +499,42 @@ future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
+    Builder& apply_atomic(const char* cell, const data_value& value);
+    template<std::ranges::range C>
+    requires std::convertible_to<std::ranges::range_value_t<C>, data_value>
+    Builder& apply_set(const char* cell, collection_apply_mode apply_mode, const C& c);
```
Unless it indeed applies only to sets.

piodul replied:
It only works with sets.

kbr-scylla commented on service/storage_service.cc:
```diff
@@ -1566,7 +1567,8 @@ future<> storage_service::raft_bootstrap(raft::server& raft_server) {
                .set("shard_count", smp::count)
-               .set("ignore_msb", _db.local().get_config().murmur3_partitioner_ignore_msb_bits());
+               .set("ignore_msb", _db.local().get_config().murmur3_partitioner_ignore_msb_bits())
+               .set("supported_features", _feature_service.supported_feature_set());
```
We should somehow deal with this duplication so we don't have to remember to update so many places.
Out of the scope of this PR.

piodul commented on db/system_keyspace.cc:
```diff
@@ -1555,6 +1555,16 @@ future<db_clock::time_point> system_keyspace::get_truncated_at(table_id cf_id) {
+    auto cdef = s.get_column_definition(name);
+    auto deserialized = cdef->type->deserialize(blob);
+    auto tokens = value_cast<set_type_impl::native_type>(deserialized);
+    return decoder(std::move(tokens));
```
Makes sense, it will be simpler that way.

piodul commented on service/storage_service.cc:
```diff
@@ -499,12 +499,42 @@ future<> storage_service::merge_topology_snapshot(raft_topology_snapshot snp) {
+class topology_node_mutation_builder
+        : public topology_mutation_builder_base<topology_node_mutation_builder> {
+
+    using builder_base = topology_mutation_builder_base<topology_node_mutation_builder>;
```
I tried it, looks like it works.

piodul commented on db/system_keyspace.cc:
```diff
@@ -1555,6 +1555,16 @@ future<db_clock::time_point> system_keyspace::get_truncated_at(table_id cf_id) {
+    auto blob = row.get_blob(name);
+    auto cdef = s.get_column_definition(name);
+    auto deserialized = cdef->type->deserialize(blob);
+    auto tokens = value_cast<set_type_impl::native_type>(deserialized);
```
No longer applies, this variable was removed.

piodul commented on db/system_keyspace.cc:
```diff
@@ -3473,6 +3473,15 @@ future<> system_keyspace::set_must_synchronize_topology(bool value) {
+    std::set<sstring> fset;
+    for (auto& f : features) {
+        auto str = value_cast<sstring>(std::move(f));
+        fset.insert(std::move(str));
```
Fair enough, removed.

piodul commented:
v3:
- `decode_set_column_with` -> `deserialize_set_column`,
- Removed `auto tokens` which used to have an unnecessarily specific name - now it doesn't exist at all,
- Elided `auto str`,
- Moved `using builder_base = ...` to the CRTP base class
